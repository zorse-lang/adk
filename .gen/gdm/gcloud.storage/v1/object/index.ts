import { GdmResource, Entity as ADKEntity } from "@zorse/adk";

export class Object extends GdmResource<ObjectComponentInputs> implements ObjectComponentOutputs {
	constructor(entity: ADKEntity, options: ObjectComponentInputs) {
		super(entity, options.name, "storage.v1.object", options);
	}
}
export interface ObjectComponentOutputs {}
export interface ObjectComponentInputs {
	readonly acl?: ObjectAccessControl[] | undefined;
	readonly bucket?: string | undefined;
	readonly cacheControl?: string | undefined;
	readonly componentCount?: number | undefined;
	readonly contentDisposition?: string | undefined;
	readonly contentEncoding?: string | undefined;
	readonly contentLanguage?: string | undefined;
	readonly contentType?: string | undefined;
	readonly crc32c?: string | undefined;
	readonly customerEncryption?: { encryptionAlgorithm?: string; keySha256?: string } | undefined;
	readonly customTime?: string | undefined;
	readonly etag?: string | undefined;
	readonly eventBasedHold?: boolean | undefined;
	readonly generation?: string | undefined;
	readonly id?: string | undefined;
	readonly kind?: string | undefined;
	readonly kmsKeyName?: string | undefined;
	readonly md5Hash?: string | undefined;
	readonly mediaLink?: string | undefined;
	readonly metadata?: { [P in string]: string } | undefined;
	readonly metageneration?: string | undefined;
	readonly name: string;
	readonly owner?: { entity?: string; entityId?: string } | undefined;
	readonly retentionExpirationTime?: string | undefined;
	readonly selfLink?: string | undefined;
	readonly size?: string | undefined;
	readonly storageClass?: string | undefined;
	readonly temporaryHold?: boolean | undefined;
	readonly timeCreated?: string | undefined;
	readonly timeDeleted?: string | undefined;
	readonly timeStorageClassUpdated?: string | undefined;
	readonly updated?: string | undefined;
	readonly type: string;
}
export interface ObjectAccessControl {
	readonly bucket?: string | undefined;
	readonly domain?: string | undefined;
	readonly email?: string | undefined;
	readonly entity?: string | undefined;
	readonly entityId?: string | undefined;
	readonly etag?: string | undefined;
	readonly generation?: string | undefined;
	readonly id?: string | undefined;
	readonly kind?: string | undefined;
	readonly object?: string | undefined;
	readonly projectTeam?: { projectNumber?: string; team?: string } | undefined;
	readonly role?: string | undefined;
	readonly selfLink?: string | undefined;
}
export default {
	Object: Object,
};
