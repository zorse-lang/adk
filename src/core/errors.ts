export class BaseError extends Error {
	constructor(...args: any[]) {
		super();
		this.name = this.constructor.name;
		this.message = this.name.replace(/([A-Z])/g, " $1").trim().toLowerCase();
		Object.assign(this, { additionalData: args });
	}
}
function ok<errT extends BaseError, condT = boolean>(
	errFactory: new (
		...args: any[]
	) => errT,
	condition: condT,
	...args: any[]
) {
	if (!condition) {
		throw new errFactory(...args);
	}
}
function no<errT extends BaseError, condT = boolean>(
	errFactory: new (
		...args: any[]
	) => errT,
	condition: condT,
	...args: any[]
) {
	if (!!condition) {
		throw new errFactory(...args);
	}
}
export const assert = { true: ok, false: no };
export namespace errors {
	export class InvalidTokenName extends BaseError {}
	export class CyclicComponentTokenPair extends BaseError {}
	export class CyclicDependencyResolver extends BaseError {}
	export class NonDirectionalSceneViolation extends BaseError {}
	export class UniDirectionalSceneViolation extends BaseError {}
	export class ComponentVerificationViolation extends BaseError {}
}
