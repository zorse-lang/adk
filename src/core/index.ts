export * from "./errors";
export * from "./object";
export * from "./resolver";
export * from "./token";
export * from "./tree";
