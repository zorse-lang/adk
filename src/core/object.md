# ADK Object Model

The Object model in ADK is inspired by Unity3D's GameObject from an API consumer
perspective, however the Object model in ADK is much more sophisticated.

The Object model is designed as a Component-Entity System. Individual Components
are able to render to multiple Scenes simultaneously. Components can also cross
reference each other across Scenes.

This design is often times referred to as a "Multi-Headed Renderer".

## Terminology

### Token

Tokens are a special data type that can be used in place of a concrete value and
with a resolver that can be used to fetch the value later, asynchronously.

Tokens can have an optional name and an optional `userData` object carrying with
them around. The Object model uses Tokens to build dependency between Components
and Scenes they are rendered into.

### Component

Components are single rendering units. Components are the lowest moving parts in
the entire Object model. Components are always attached to an Entity.  
No Assumptions are ever made about Components except they subclass the abstract
class `Component` and implement the `render()` and `verify()` method.

Components render into an opaque object with type `any`.

### Entity

Entities are logical groupings of Components for end user application.  
Entities orchestrate their child Components to render into multiple Scenes.

### Scene

Scenes are what Entities render (flatten) their Components into. Scenes are made
of Views. Scenes may be of three types:

- `Scene.Type.Nondirectional`: Components in this scene can't reference other
  Components in other Scenes.
- `Scene.Type.Unidirectional`: Components in this scene can only reference
  Components in other Scenes, but other Scenes can't reference them.
- `Scene.Type.Bidirectional`: Components in this scene can reference other
  Components in other Scenes, and other Scenes can reference them as well.

### View

Views are the lowest level of rendering unit in the Object model. Views are what
Components in each scene eventually render into.

### Composition

Compositions are the final output of the Object model. Compositions are ordered
collection of independent Views that can be consumed by the end user.

### System

The System is the entry point to the Object model. The System is responsible for
orchestrating the entire rendering process.

## Render Loop

```mermaid
stateDiagram
    [*] --> System.compose(): Entities, Components, and Scenes are inputs to System
    System.compose() --> Entity.update(): Begin Entity level rendering
    Entity.update() --> Entity.render(): Resolve Tokens if possible
    Entity.render() --> Scene.accepts(Component): Flatten Entity-Component tree into Scenes
    Scene.accepts(Component) --> Scene.replace(Token): Let Scenes handle unresolved Tokens
    Scene.replace(Token) --> View.render(): Create a "shadow" Component with replaced Tokens
    View.render() --> Component.update(): Begin Component level rendering
    Component.update() --> Component.verify(): Resolve replaced Tokens
    Component.verify() --> Component.render(): Render the ShadowComponent
    Component.render() --> Scene.cluster(): Cluster Views based on Scene limits
    Scene.cluster() --> Composition: Wrap Views for user consumption
    Composition --> [*]: Independent Views as outputs
```
