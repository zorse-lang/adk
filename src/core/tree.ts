/** A general purpose Tree implementation */
export class Tree {
	/** The name of the tree. Defaults to type name of the class plus count of same type of Tree in the parent Tree. */
	private readonly _name?: string;
	/** The parent Tree. Lack of a parent indicates this is a root tree */
	public readonly parent?: Tree;
	// Children are Trees themselves (subtrees)
	private readonly _children: Set<Tree> = new Set();
	/**
	 * Creates a new Tree.
	 * @param parent The parent Tree if any.
	 * @param name The name of the tree.
	 */
	public constructor(parent?: Tree, name?: string) {
		this.parent = parent;
		this._name =
			name ||
			(() => {
				const type = this.constructor.name;
				const count = [...this.parent._children].filter((child) => child.constructor.name === type).length;
				return `${type}${count === 0 ? "" : `${count}`}`;
			})();
		if (this.parent?._children) {
			this.parent._children.add(this);
		}
	}
	/** Returns root of the tree */
	public root(): Tree {
		return this.parent ? this.parent.root() : this;
	}
	/** Returns the path of the tree from root to this tree. */
	public path(opts: Tree.PathOptions = {}): string {
		opts = { ...TREE_DEFAULT_OPTIONS, ...opts };
		const _path = this.parent ? `${this.parent.path(opts)}${opts.sep}${this._name}` : opts.noroot ? "" : this._name;
		const paths = _path.split(opts.sep).filter(Boolean);
		const _maybeReversed = opts.reverse ? paths.reverse().join(opts.sep) : paths.join(opts.sep);
		return _maybeReversed;
	}
	/** Returns the children of the tree. */
	public children(): Tree[] {
		return [...this._children];
	}
}

export namespace Tree {
	export interface PathOptions {
		/** The separator to use between each tree name. Defaults to `/` */
		readonly sep?: string;
		/** Whether to include the root tree name in the path. Defaults to `false` */
		readonly noroot?: boolean;
		/** Whether to reverse the path. Defaults to `false` */
		readonly reverse?: boolean;
	}
}

const TREE_DEFAULT_OPTIONS: Tree.PathOptions = {
	sep: "/",
	noroot: false,
	reverse: false,
};
