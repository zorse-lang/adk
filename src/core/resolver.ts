import { assert, errors } from "@zorse/adk/core";

/**
 * A 2D basic dependency resolver.
 * You can be creative and turn a 2D dependency resolver into a 3D, 4D, 5D, etc. dependency resolvers
 * by grouping dependencies into a single object. That's how the {@link core.System:class} composes
 * {@link core.Scene:class}s and {@link core.Entity:class}es.
 */
export class Resolver<T = any> {
	public readonly members = new Map<T, Set<T>>();
	/**
	 * Solves the dependency graph and returns a list of clusters.
	 * @returns A list of clusters. Each cluster is a list of members that are dependent on each other.
	 */
	public resolve(): T[][] {
		const clusters: Array<Set<T>> = [];
		const all = new Set<T>();
		this.members.forEach((dependencies, dependent) => {
			all.add(dependent);
			dependencies.forEach((dependency) => {
				all.add(dependency);
			});
		});
		const found = (item: T): boolean => {
			return clusters.some((cluster) => cluster.has(item));
		};
		for (const item of all) {
			if (!found(item)) {
				const newCluster = new Set<T>();
				newCluster.add(item);
				const deps = this.getDependencies(item);
				deps.forEach((dep) => {
					if (!found(dep)) {
						newCluster.add(dep);
					}
				});
				clusters.push(newCluster);
			}
		}
		const resolved = clusters.map((cluster) => Array.from(cluster));
		return resolved;
	}
	/**
	 * Adds an isolated member to the dependency graph. A member without any dependencies.
	 * @param member The member to add.
	 */
	public addIsolated(member: T): void {
		if (!this.members.has(member)) {
			this.members.set(member, new Set());
		}
	}
	/**
	 * Adds a dependent pair to the dependency graph.
	 * @param dependent The dependent member.
	 * @param dependency The dependency member.
	 */
	public addDependency(dependent: T, ...dependency: T[]): void {
		this.addIsolated(dependent);
		dependency.forEach((dep) => {
			this.addIsolated(dep);
			assert.false(errors.CyclicDependencyResolver, this.getDependencies(dep).has(dependent), dependent, dependency);
			this.members.get(dependent).add(dep);
		});
	}
	/**
	 * Recursively gets all dependencies of a single member.
	 * @param dependent The dependent member.
	 */
	public getDependencies(dependent: T): Set<T> {
		const out = new Set<T>();
		this._getDependencies(dependent, out);
		return out;
	}
	private _getDependencies(dependent: T, out: Set<T>): void {
		const resolve = (dep: T) => {
			if (this.members.has(dep)) {
				this.members.get(dep).forEach((dependency) => {
					if (!out.has(dependency)) {
						out.add(dependency);
						resolve(dependency);
					}
				});
			}
		};
		resolve(dependent);
	}
}
