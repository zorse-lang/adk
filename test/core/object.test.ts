import * as ECS from "@zorse/adk/core/object";

import { Token, errors } from "@zorse/adk/core";

class DebugScene extends ECS.Scene {
	filter(component: ECS.Component): boolean {
		return component instanceof DebugComponent;
	}
}
class DebugComponent extends ECS.Component {
	public someFictionalProperty?: string;
	public moreFictionalProperty?: string;
	constructor(parent: ECS.Entity, public readonly properties: any = {}) {
		super(parent);
	}
	[ECS.Symbols.ComponentRender](output: any) {
		if (!output.Components) {
			output.Components = [];
		}
		output.Components.push(this[ECS.Symbols.ComponentHandle].serialize());
	}
}

describe("Object model tests", () => {
	it("should be able to render basic components", async () => {
		const system = new ECS.System();
		const entity = new ECS.Entity(system, "OptionalName");
		new DebugComponent(entity, { hello: "world" });
		new DebugComponent(entity, { hello: "example" });
		const otherEntity = new ECS.Entity(system);
		new DebugComponent(otherEntity, { hello: "world" });
		new DebugComponent(otherEntity, { hello: "sample" });
		const scene = new DebugScene(system);
		await system.compose();
		expect(scene.gizmos()).toEqual({
			Components: [
				{ properties: { hello: "world" } },
				{ properties: { hello: "example" } },
				{ properties: { hello: "world" } },
				{ properties: { hello: "sample" } },
			],
		});
	});

	it("should be able to render components with tokens", async () => {
		const system = new ECS.System();
		const entity = new ECS.Entity(system);
		new DebugComponent(entity, { token: new Token({ name: "TokenName" }) });
		const scene = new DebugScene(system);
		await system.compose();
		expect(scene.gizmos()).toEqual({
			Components: [{ properties: { token: "@@{TokenName}@@" } }],
		});
	});

	it("should be able to resolve components with tokens", async () => {
		const system = new ECS.System();
		const entity = new ECS.Entity(system);
		new DebugComponent(entity, {
			token: new Token({
				name: "TokenWithResolver",
				resolver: () => "some value",
			}),
		});
		new DebugComponent(entity, {
			token: new Token({ name: "unresolved" }),
		});
		const scene = new DebugScene(system);
		await system.compose();
		expect(scene.gizmos()).toEqual({
			Components: [{ properties: { token: "some value" } }, { properties: { token: "@@{unresolved}@@" } }],
		});
	});

	it("should report path to a nested Entity", async () => {
		const system = new ECS.System();
		const entity = new ECS.Entity(system);
		const nested = new ECS.Entity(entity);
		new DebugScene(system);
		await system.compose();
		expect(nested.path()).toEqual("Entity/Entity");
	});

	it("should report path to a nested Custom Entity", async () => {
		class CustomEntity1 extends ECS.Entity {}
		class CustomEntity2 extends ECS.Entity {}
		const system = new ECS.System();
		const entity = new CustomEntity1(system);
		const nested = new CustomEntity2(entity);
		new DebugScene(system);
		await system.compose();
		expect(nested.path()).toEqual("CustomEntity1/CustomEntity2");
	});

	it("should return Tokens if Components accesses are undefined", async () => {
		const system = new ECS.System();
		const entity = new ECS.Entity(system);
		const component1 = new DebugComponent(entity, { existing: 235 });
		const value1 = (component1 as any).random;
		expect(value1).toBeInstanceOf(Token);
		expect(value1.name()).toBe('Entity/DebugComponent["random"]');
		const resolved1 = await (value1 as Token).resolve();
		expect(resolved1).toBe(value1);
		const component2 = new DebugComponent(entity, { other: { value: "example" } });
		const value2 = (component2 as any).other.random[0];
		expect(value2.name()).toBe('Entity/DebugComponent2["other"]["random"]["0"]');
		expect(value2).toBeInstanceOf(Token);
		const resolved2 = await (value2 as Token).resolve();
		expect(`${resolved2}`).toBe(`${value2}`);
		expect(component1.properties.existing).toBe(235);
	});

	it("should throw on cyclic scenes", async () => {
		class DebugComponent1 extends DebugComponent {}
		class DebugComponent2 extends DebugComponent {}
		class Scene1 extends ECS.Scene {
			filter(component: ECS.Component): boolean {
				return component instanceof DebugComponent1;
			}
		}
		class Scene2 extends ECS.Scene {
			filter(component: ECS.Component): boolean {
				return component instanceof DebugComponent2;
			}
		}
		const system = new ECS.System();
		const entity = new ECS.Entity(system);
		const component1 = new DebugComponent1(entity);
		const component2 = new DebugComponent2(entity, { hello: component1.someFictionalProperty });
		component1.properties.hello = component2.someFictionalProperty;
		new Scene1(system);
		new Scene2(system);
		expect(system.compose()).rejects.toThrow(errors.CyclicDependencyResolver);
	});

	it("should be able to compose Bidirectional scenes", async () => {
		class Scene1 extends ECS.Scene {
			filter(component: ECS.Component): boolean {
				return (
					component instanceof Component_A ||
					component instanceof Component_B ||
					component instanceof Component_C ||
					component instanceof Component_D ||
					component instanceof Component_E ||
					component instanceof Component_F ||
					component instanceof Component_G ||
					component instanceof Component_H ||
					component instanceof Component_I ||
					component instanceof Component_J
				);
			}
		}

		class Scene2 extends ECS.Scene {
			filter(component: ECS.Component): boolean {
				return (
					component instanceof Component_a ||
					component instanceof Component_b ||
					component instanceof Component_c ||
					component instanceof Component_d ||
					component instanceof Component_e ||
					component instanceof Component_f ||
					component instanceof Component_g ||
					component instanceof Component_h ||
					component instanceof Component_i ||
					component instanceof Component_j
				);
			}
		}

		class Scene3 extends ECS.Scene {
			filter(component: ECS.Component): boolean {
				return (
					component instanceof Component_0 ||
					component instanceof Component_1 ||
					component instanceof Component_2 ||
					component instanceof Component_3 ||
					component instanceof Component_4 ||
					component instanceof Component_5 ||
					component instanceof Component_6 ||
					component instanceof Component_7 ||
					component instanceof Component_8 ||
					component instanceof Component_9
				);
			}
		}

		class Entity_ABC extends ECS.Entity {}
		class Entity_EFGHI extends ECS.Entity {}
		class Entity_bcdef extends ECS.Entity {}
		class Entity_012 extends ECS.Entity {}
		class Entity_789 extends ECS.Entity {}

		class Component_A extends DebugComponent {}
		class Component_B extends DebugComponent {}
		class Component_C extends DebugComponent {}
		class Component_D extends DebugComponent {}
		class Component_E extends DebugComponent {}
		class Component_F extends DebugComponent {}
		class Component_G extends DebugComponent {}
		class Component_H extends DebugComponent {}
		class Component_I extends DebugComponent {}
		class Component_J extends DebugComponent {}

		class Component_a extends DebugComponent {}
		class Component_b extends DebugComponent {}
		class Component_c extends DebugComponent {}
		class Component_d extends DebugComponent {}
		class Component_e extends DebugComponent {}
		class Component_f extends DebugComponent {}
		class Component_g extends DebugComponent {}
		class Component_h extends DebugComponent {}
		class Component_i extends DebugComponent {}
		class Component_j extends DebugComponent {}

		class Component_0 extends DebugComponent {}
		class Component_1 extends DebugComponent {}
		class Component_2 extends DebugComponent {}
		class Component_3 extends DebugComponent {}
		class Component_4 extends DebugComponent {}
		class Component_5 extends DebugComponent {}
		class Component_6 extends DebugComponent {}
		class Component_7 extends DebugComponent {}
		class Component_8 extends DebugComponent {}
		class Component_9 extends DebugComponent {}

		const system = new ECS.System();
		const entity_ABC = new Entity_ABC(system);
		const c_A = new Component_A(entity_ABC);
		const c_C = new Component_C(entity_ABC);
		const c_B = new Component_B(entity_ABC, { hello: c_A.someFictionalProperty, world: c_C.someFictionalProperty });
		new Component_D(system.root);
		const entity_EFGHI = new Entity_EFGHI(system);
		const c_E = new Component_E(entity_EFGHI);
		const c_H = new Component_H(entity_EFGHI);
		const c_I = new Component_I(entity_EFGHI);
		const c_G = new Component_G(entity_EFGHI, { hello: c_I.someFictionalProperty, world: c_H.someFictionalProperty });
		const c_F = new Component_F(entity_EFGHI, { hello: c_E.someFictionalProperty, world: c_G.someFictionalProperty });
		new Component_J(system.root);
		const c_a = new Component_a(system.root);
		const entity_bcdef = new Entity_bcdef(system);
		const c_f = new Component_f(entity_bcdef);
		const c_e = new Component_e(entity_bcdef, { hello: c_f.someFictionalProperty });
		const c_d = new Component_d(entity_bcdef, { hello: c_e.someFictionalProperty });
		const c_c = new Component_c(entity_bcdef, { hello: c_d.someFictionalProperty });
		const c_b = new Component_b(entity_bcdef, { hello: c_c.someFictionalProperty });
		new Component_g(system.root);
		new Component_h(system.root);
		new Component_i(system.root);
		const c_j = new Component_j(system.root);
		const entity_012 = new Entity_012(system);
		const c_1 = new Component_1(entity_012);
		const c_2 = new Component_2(entity_012);
		const c_0 = new Component_0(entity_012, { hello: c_1.someFictionalProperty, world: c_2.someFictionalProperty });
		new Component_3(system.root);
		new Component_4(system.root);
		new Component_5(system.root);
		new Component_6(system.root);
		const entity_789 = new Entity_789(system);
		const c_7 = new Component_7(entity_789);
		const c_8 = new Component_8(entity_789, { hello: c_7.someFictionalProperty });
		const c_9 = new Component_9(entity_789, { hello: c_8.moreFictionalProperty });

		// row 1: A-B-C D E-F-G-H-I J
		//          |       ^
		//          v       |
		// row 2: a b-c-d-e-f g h i j
		//        |                 |
		//        v                 v
		// row 3: 0-1-2 3 4 5 6 7-8-9

		c_B.someFictionalProperty = c_b.moreFictionalProperty;
		c_a.someFictionalProperty = c_0.moreFictionalProperty;
		c_f.moreFictionalProperty = c_F.someFictionalProperty;
		c_j.moreFictionalProperty = c_9.someFictionalProperty;

		const scene1 = new Scene1(system);
		const scene2 = new Scene2(system);
		const scene3 = new Scene3(system);

		const composition = await system.compose();
		const expected = [
			{
				Components: [
					{ properties: { hello: '@@{Entity_789/Component_83["moreFictionalProperty"]}@@' } },
					{ properties: { hello: '@@{Entity_789/Component_73["someFictionalProperty"]}@@' } },
					{ properties: {} },
					{ properties: {} },
					{ properties: {} },
					{ properties: {} },
					{ properties: {} },
				],
			},
			{
				Components: [
					{ properties: {} },
					{ properties: {} },
					{ properties: {} },
					{ properties: {}, moreFictionalProperty: '@@{Entity_789/Component_93["someFictionalProperty"]}@@' },
				],
			},
			{
				Components: [
					{
						properties: {
							hello: '@@{Entity_012/Component_13["someFictionalProperty"]}@@',
							world: '@@{Entity_012/Component_23["someFictionalProperty"]}@@',
						},
					},
					{ properties: {} },
					{ properties: {} },
				],
			},
			{
				Components: [
					{ properties: {}, someFictionalProperty: '@@{Entity_012/Component_03["moreFictionalProperty"]}@@' },
				],
			},
			{
				Components: [
					{
						properties: {
							hello: '@@{Entity_EFGHI/Component_E5["someFictionalProperty"]}@@',
							world: '@@{Entity_EFGHI/Component_G5["someFictionalProperty"]}@@',
						},
					},
					{ properties: {} },
					{
						properties: {
							hello: '@@{Entity_EFGHI/Component_I5["someFictionalProperty"]}@@',
							world: '@@{Entity_EFGHI/Component_H5["someFictionalProperty"]}@@',
						},
					},
					{ properties: {} },
					{ properties: {} },
				],
			},
			{
				Components: [
					{ properties: { hello: '@@{Entity_bcdef/Component_c5["someFictionalProperty"]}@@' } },
					{ properties: { hello: '@@{Entity_bcdef/Component_d5["someFictionalProperty"]}@@' } },
					{ properties: { hello: '@@{Entity_bcdef/Component_e5["someFictionalProperty"]}@@' } },
					{ properties: { hello: '@@{Entity_bcdef/Component_f5["someFictionalProperty"]}@@' } },
					{ properties: {}, moreFictionalProperty: '@@{Entity_EFGHI/Component_F5["someFictionalProperty"]}@@' },
				],
			},
			{
				Components: [
					{ properties: {} },
					{ properties: {} },
					{ properties: {} },
					{ properties: {} },
					{
						properties: {
							hello: '@@{Entity_ABC/Component_A3["someFictionalProperty"]}@@',
							world: '@@{Entity_ABC/Component_C3["someFictionalProperty"]}@@',
						},
						someFictionalProperty: '@@{Entity_bcdef/Component_b5["moreFictionalProperty"]}@@',
					},
				],
			},
		];
		expect(composition.scenes).toEqual([scene3, scene2, scene1]);
		expect(composition.gizmos()).toEqual(expected);
	});
});
