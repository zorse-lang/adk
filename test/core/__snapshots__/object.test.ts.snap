// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Object model tests should be able to compose Bidirectional scenes 1`] = `
[
  {
    "Components": [
      {
        "properties": {
          "hello": "@@{Entity_789/Component_8/moreFictionalProperty}@@",
        },
      },
      {
        "properties": {
          "hello": "@@{Entity_789/Component_7/someFictionalProperty}@@",
        },
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
    ],
  },
  {
    "Components": [
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "moreFictionalProperty": "@@{Entity_789/Component_9/someFictionalProperty}@@",
        "properties": {},
      },
    ],
  },
  {
    "Components": [
      {
        "properties": {
          "hello": "@@{Entity_012/Component_1/someFictionalProperty}@@",
          "world": "@@{Entity_012/Component_2/someFictionalProperty}@@",
        },
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
    ],
  },
  {
    "Components": [
      {
        "properties": {},
        "someFictionalProperty": "@@{Entity_012/Component_0/moreFictionalProperty}@@",
      },
    ],
  },
  {
    "Components": [
      {
        "properties": {
          "hello": "@@{Entity_EFGHI/Component_E/someFictionalProperty}@@",
          "world": "@@{Entity_EFGHI/Component_G/someFictionalProperty}@@",
        },
      },
      {
        "properties": {},
      },
      {
        "properties": {
          "hello": "@@{Entity_EFGHI/Component_I/someFictionalProperty}@@",
          "world": "@@{Entity_EFGHI/Component_H/someFictionalProperty}@@",
        },
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
    ],
  },
  {
    "Components": [
      {
        "properties": {
          "hello": "@@{Entity_bcdef/Component_c/someFictionalProperty}@@",
        },
      },
      {
        "properties": {
          "hello": "@@{Entity_bcdef/Component_d/someFictionalProperty}@@",
        },
      },
      {
        "properties": {
          "hello": "@@{Entity_bcdef/Component_e/someFictionalProperty}@@",
        },
      },
      {
        "properties": {
          "hello": "@@{Entity_bcdef/Component_f/someFictionalProperty}@@",
        },
      },
      {
        "moreFictionalProperty": "@@{Entity_EFGHI/Component_F/someFictionalProperty}@@",
        "properties": {},
      },
    ],
  },
  {
    "Components": [
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {},
      },
      {
        "properties": {
          "hello": "@@{Entity_ABC/Component_A/someFictionalProperty}@@",
          "world": "@@{Entity_ABC/Component_C/someFictionalProperty}@@",
        },
        "someFictionalProperty": "@@{Entity_bcdef/Component_b/moreFictionalProperty}@@",
      },
    ],
  },
]
`;

exports[`Object model tests should be able to render basic components 1`] = `
{
  "Components": [
    {
      "properties": {
        "hello": "world",
      },
    },
    {
      "properties": {
        "hello": "example",
      },
    },
    {
      "properties": {
        "hello": "world",
      },
    },
    {
      "properties": {
        "hello": "sample",
      },
    },
  ],
}
`;

exports[`Object model tests should be able to render components with string embedded tokens 1`] = `
[
  {
    "Components": [
      {
        "properties": {
          "token": "prefix-@@{TokenName}@@",
        },
      },
    ],
  },
]
`;

exports[`Object model tests should be able to render components with tokens 1`] = `
{
  "Components": [
    {
      "properties": {
        "token": "@@{TokenName}@@",
      },
    },
  ],
}
`;

exports[`Object model tests should be able to resolve components with tokens 1`] = `
{
  "Components": [
    {
      "properties": {
        "token": "some value",
      },
    },
    {
      "properties": {
        "token": "@@{unresolved}@@",
      },
    },
  ],
}
`;
